cmsl1t/hist/binning.py

```diff
@@ -0,0 +1,125 @@
+import bisect
+from exceptions import KeyError
+from copy import deepcopy
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class Base():
+    overflow = "overflow"
+    underflow = "underflow"
+    everything = "everything"
+
+    def __init__(self, label, n_bins, use_everything_bin=False):
+        self.use_everything_bin = use_everything_bin
+        self.label = label
+        self.n_bins = n_bins
+        keys = range(self.n_bins) + [self.overflow, self.underflow]
+        if use_everything_bin:
+            keys += [self.everything]
+        self.values = {}
+        for i in keys:
+            self.values[i] = None
+
+    def set_all_values(self, value):
+        for key in self.values:
+            self.values[key] = deepcopy(value)
+
+    def set_value(self, bin, value):
+        self.values[bin] = value
+
+    def __len__(self):
+        return self.n_bins
+
+    def get_bin_contents(self, bin_index):
+        contents = self.values.get(bin_index, "DoesntExist")
+        if contents is "DoesntExist":
+            msg = "Cannot find bin for index, {0}, for binning called '{1}'"
+            logger.error(msg.format(bin_index, self.label))
+            raise KeyError(bin_index)
+        return contents
+
+    def find_all_bins(self, value):
+        bins = self.find_bins(value)
+        if self.use_everything_bin:
+            bins.append(self.everything)
+        return bins
+
+    def __getitem__(self, value):
+        bins = self.find_all_bins(value)
+        return [self.get_bin_contents(i) for i in bins]
+
+    def __iter__(self):
+        for i in range(self.n_bins):
+            yield i
+
+    def iter_all(self):
+        for i in self.values:
+            yield i
+
+
+class Sorted(Base):
+    import bisect
+
+    def __init__(self, bin_edges, label=None, use_everything_bin=False):
```

Comment:
> why swap it around? 

I'm not really swapping them around, since what's passed through is not the exact binning, but just the number of bins. There's no harm in making the order agree though, so I've swapped things round in the base class init. 

> What are the cases where label is not needed?

Originally it was just a decoration, didn't really need it, but thought it might be useful.
Now (in commits after these changes) it's become mandatory, so I've removed the default value.